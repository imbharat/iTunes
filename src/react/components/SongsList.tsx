import React from 'react'
import InfiniteScroll from 'react-infinite-scroll-component';
import { CircularProgress } from '@mui/material';
import { useAppDispatch, useAppSelector } from '../../redux/ReduxHooks';
import { updateFilterAndLoadSongs } from '../../redux/Songs/SongsActions';
import { Song as SongObject, Songs } from '../../redux/Songs/Types'
import Song from './Song'

function SongsList({ initialData } : { initialData: unknown }) {
  const songsState = useAppSelector((state: Songs) => state);
  
  const dispatch = useAppDispatch();
  
  const fetchNext = () => {
    dispatch(updateFilterAndLoadSongs({
      ...filter,
      page: filter.page + 1
    }, false))
  }
  
  let songs: SongObject[] = [];
  if(songsState.songs?.length){
    songs = songsState.songs; 
  }
  else {
    songs = JSON.parse(initialData as string).data;
  }

  const {
    isLast,
    filter
  } = songsState
  
  return (
    <div>
      <h2 style={{textAlign: 'center', marginBottom: 0}}><b>TOP RESULTS</b></h2>
      <InfiniteScroll
        style={{
          display: 'flex',
          flexWrap: 'wrap'
        }}
        dataLength={songs.length}
        next={fetchNext}
        hasMore={!isLast}
        loader={<CircularProgress />}
        endMessage={
          isLast && <h2 style={{ textAlign: 'center', width: '100%'}}>
            That's all we have!
          </h2>
        }
      >
        {
          songs.map((song, index) => <Song key={`${ index }_${song.trackId}`} song={song}/>)
        }
      </InfiniteScroll>
    </div>
  )
}

export default SongsList