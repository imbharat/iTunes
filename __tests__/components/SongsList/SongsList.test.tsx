import React from 'react';
import renderer from 'react-test-renderer';
import SongsList from '../../../src/react/components/SongsList';
import { initialData } from './SongsListTestData';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';

const middlewares = [thunk]
const mockStore = configureStore(middlewares);

describe("SongsList Component Tests", () => {
  let store: any;
  beforeEach(() => {
    store = mockStore({
      loading: false,
      songs: [],
      filter: {
          term: '',
          country: '',
          media: '',
          entity: '',
          attribute: '',
          limit: 50,
          lang: '',
          page: 0
      },
      isLast: true,
      error: ''
    });
  });

  it('songs list component initial rendering', () => {
    //arrange
    const testData = initialData;
    //act
    const tree = renderer.create(
      <Provider store={store}>
        <SongsList initialData={testData}/>
      </Provider>).toJSON();
    //assert
    expect(tree).toMatchSnapshot();
  });
});