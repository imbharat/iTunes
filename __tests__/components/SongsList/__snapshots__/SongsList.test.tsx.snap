// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongsList Component Tests songs list component initial rendering 1`] = `
<div>
  <h2
    style={
      {
        "marginBottom": 0,
        "textAlign": "center",
      }
    }
  >
    <b>
      TOP RESULTS
    </b>
  </h2>
  <div
    className="infinite-scroll-component__outerdiv"
    style={{}}
  >
    <div
      className="infinite-scroll-component "
      style={
        {
          "WebkitOverflowScrolling": "touch",
          "display": "flex",
          "flexWrap": "wrap",
          "height": "auto",
          "overflow": "auto",
        }
      }
    >
      <div
        className="sc-bczRLJ gBgBPG"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          onClick={[Function]}
        >
          <img
            alt="G9"
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            src="https://is4-ssl.mzstatic.com/image/thumb/Music124/v4/88/aa/ed/88aaed59-2c30-27d4-7468-4b95ef5ee7bf/mzm.iwrtpfbi.jpg/100x100bb.jpg"
          />
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom trackN css-1vw6mcs-MuiTypography-root"
            >
              G9
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom collectionN css-18k87ye-MuiTypography-root"
            >
              Space Ship One
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom artistN css-18k87ye-MuiTypography-root"
            >
              -by: Paul Gilbert
            </div>
          </div>
        </div>
      </div>
      <div
        className="sc-bczRLJ gBgBPG"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          onClick={[Function]}
        >
          <img
            alt="Hinahanap Ng Puso"
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            src="https://is1-ssl.mzstatic.com/image/thumb/Music42/v4/8c/ef/62/8cef6284-2eb7-a7fa-2d1e-2075382cbfee/cover.jpg/100x100bb.jpg"
          />
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom trackN css-1vw6mcs-MuiTypography-root"
            >
              Hinahanap Ng Puso
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom collectionN css-18k87ye-MuiTypography-root"
            >
              G9
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom artistN css-18k87ye-MuiTypography-root"
            >
              -by: Gloc-9 & Hannah Romawac
            </div>
          </div>
        </div>
      </div>
      <div
        className="sc-bczRLJ gBgBPG"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          onClick={[Function]}
        >
          <img
            alt="Simpleng Tao"
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            src="https://is1-ssl.mzstatic.com/image/thumb/Music42/v4/8c/ef/62/8cef6284-2eb7-a7fa-2d1e-2075382cbfee/cover.jpg/100x100bb.jpg"
          />
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom trackN css-1vw6mcs-MuiTypography-root"
            >
              Simpleng Tao
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom collectionN css-18k87ye-MuiTypography-root"
            >
              G9
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom artistN css-18k87ye-MuiTypography-root"
            >
              -by: Gloc-9
            </div>
          </div>
        </div>
      </div>
      <div
        className="sc-bczRLJ gBgBPG"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          onClick={[Function]}
        >
          <img
            alt="Bakit"
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            src="https://is1-ssl.mzstatic.com/image/thumb/Music42/v4/8c/ef/62/8cef6284-2eb7-a7fa-2d1e-2075382cbfee/cover.jpg/100x100bb.jpg"
          />
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom trackN css-1vw6mcs-MuiTypography-root"
            >
              Bakit
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom collectionN css-18k87ye-MuiTypography-root"
            >
              G9
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom artistN css-18k87ye-MuiTypography-root"
            >
              -by: Gloc-9 & Cookie Chua
            </div>
          </div>
        </div>
      </div>
      <div
        className="sc-bczRLJ gBgBPG"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          onClick={[Function]}
        >
          <img
            alt="Laklak"
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            src="https://is1-ssl.mzstatic.com/image/thumb/Music42/v4/8c/ef/62/8cef6284-2eb7-a7fa-2d1e-2075382cbfee/cover.jpg/100x100bb.jpg"
          />
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom trackN css-1vw6mcs-MuiTypography-root"
            >
              Laklak
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom collectionN css-18k87ye-MuiTypography-root"
            >
              G9
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom artistN css-18k87ye-MuiTypography-root"
            >
              -by: Gloc-9 & Dong Abay
            </div>
          </div>
        </div>
      </div>
      <div
        className="sc-bczRLJ gBgBPG"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          onClick={[Function]}
        >
          <img
            alt="Sikat Na Si Pepe"
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            src="https://is1-ssl.mzstatic.com/image/thumb/Music42/v4/8c/ef/62/8cef6284-2eb7-a7fa-2d1e-2075382cbfee/cover.jpg/100x100bb.jpg"
          />
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom trackN css-1vw6mcs-MuiTypography-root"
            >
              Sikat Na Si Pepe
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom collectionN css-18k87ye-MuiTypography-root"
            >
              G9
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom artistN css-18k87ye-MuiTypography-root"
            >
              -by: Gloc-9
            </div>
          </div>
        </div>
      </div>
      <div
        className="sc-bczRLJ gBgBPG"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          onClick={[Function]}
        >
          <img
            alt="Jologs"
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            src="https://is1-ssl.mzstatic.com/image/thumb/Music42/v4/8c/ef/62/8cef6284-2eb7-a7fa-2d1e-2075382cbfee/cover.jpg/100x100bb.jpg"
          />
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom trackN css-1vw6mcs-MuiTypography-root"
            >
              Jologs
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom collectionN css-18k87ye-MuiTypography-root"
            >
              G9
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom artistN css-18k87ye-MuiTypography-root"
            >
              -by: Gloc-9
            </div>
          </div>
        </div>
      </div>
      <div
        className="sc-bczRLJ gBgBPG"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          onClick={[Function]}
        >
          <img
            alt="Sayang"
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            src="https://is1-ssl.mzstatic.com/image/thumb/Music42/v4/8c/ef/62/8cef6284-2eb7-a7fa-2d1e-2075382cbfee/cover.jpg/100x100bb.jpg"
          />
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom trackN css-1vw6mcs-MuiTypography-root"
            >
              Sayang
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom collectionN css-18k87ye-MuiTypography-root"
            >
              G9
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom artistN css-18k87ye-MuiTypography-root"
            >
              -by: Gloc-9
            </div>
          </div>
        </div>
      </div>
      <div
        className="sc-bczRLJ gBgBPG"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          onClick={[Function]}
        >
          <img
            alt="Si Raul"
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            src="https://is1-ssl.mzstatic.com/image/thumb/Music42/v4/8c/ef/62/8cef6284-2eb7-a7fa-2d1e-2075382cbfee/cover.jpg/100x100bb.jpg"
          />
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom trackN css-1vw6mcs-MuiTypography-root"
            >
              Si Raul
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom collectionN css-18k87ye-MuiTypography-root"
            >
              G9
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom artistN css-18k87ye-MuiTypography-root"
            >
              -by: Gloc-9
            </div>
          </div>
        </div>
      </div>
      <div
        className="sc-bczRLJ gBgBPG"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          onClick={[Function]}
        >
          <img
            alt="Ayoko Na Sana"
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            src="https://is1-ssl.mzstatic.com/image/thumb/Music42/v4/8c/ef/62/8cef6284-2eb7-a7fa-2d1e-2075382cbfee/cover.jpg/100x100bb.jpg"
          />
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom trackN css-1vw6mcs-MuiTypography-root"
            >
              Ayoko Na Sana
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom collectionN css-18k87ye-MuiTypography-root"
            >
              G9
            </div>
            <div
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom artistN css-18k87ye-MuiTypography-root"
            >
              -by: Gloc-9
            </div>
          </div>
        </div>
      </div>
      <h2
        style={
          {
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        That's all we have!
      </h2>
    </div>
  </div>
</div>
`;
